File: rest/cursor_api/database/queries.py
"""Database query functions for reading chat data"""

import json
import sqlite3
from typing import List, Dict, Any, Optional, Tuple


def get_all_chat_ids(conn: sqlite3.Connection) -> List[str]:
    """Get all chat IDs from database
    
    Args:
        conn: SQLite connection
        
    Returns:
        List of chat IDs
    """
    cursor = conn.cursor()
    cursor.execute("""
        SELECT key FROM cursorDiskKV 
        WHERE key LIKE 'composerData:%'
    """)
    
    chat_ids = []
    for (key,) in cursor.fetchall():
        chat_id = key.replace('composerData:', '')
        chat_ids.append(chat_id)
    
    return chat_ids


def get_chat_metadata(conn: sqlite3.Connection, chat_id: str) -> Optional[Dict[str, Any]]:
    """Get metadata for a specific chat
    
    Args:
        conn: SQLite connection
        chat_id: Chat UUID
        
    Returns:
        Chat metadata dictionary or None if not found
    """
    cursor = conn.cursor()
    cursor.execute(
        "SELECT value FROM cursorDiskKV WHERE key = ?",
        (f'composerData:{chat_id}',)
    )
    
    row = cursor.fetchone()
    if not row:
        return None
    
    return json.loads(row[0])


def get_chat_messages(
    conn: sqlite3.Connection,
    chat_id: str
) -> List[Tuple[str, Dict[str, Any]]]:
    """Get all messages for a specific chat
    
    Args:
        conn: SQLite connection
        chat_id: Chat UUID
        
    Returns:
        List of tuples (bubble_id, bubble_data)
    """
    cursor = conn.cursor()
    cursor.execute("""
        SELECT key, value 
        FROM cursorDiskKV 
        WHERE key LIKE ?
        ORDER BY key
    """, (f'bubbleId:{chat_id}:%',))
    
    messages = []
    for key, value_blob in cursor.fetchall():
        bubble_id = key.split(':')[-1]
        bubble_data = json.loads(value_blob)
        messages.append((bubble_id, bubble_data))
    
    return messages


def count_chats(conn: sqlite3.Connection, include_archived: bool = True) -> int:
    """Count total number of chats
    
    Args:
        conn: SQLite connection
        include_archived: Whether to include archived chats
        
    Returns:
        Total count of chats
    """
    cursor = conn.cursor()
    
    if include_archived:
        cursor.execute("""
            SELECT COUNT(*) FROM cursorDiskKV 
            WHERE key LIKE 'composerData:%'
        """)
    else:
        # Need to check each chat's metadata
        cursor.execute("""
            SELECT value FROM cursorDiskKV 
            WHERE key LIKE 'composerData:%'
        """)
        count = 0
        for (value_blob,) in cursor.fetchall():
            metadata = json.loads(value_blob)
            if not metadata.get('isArchived', False):
                count += 1
        return count
    
    return cursor.fetchone()[0]

File: rest-rust/src/db/queries.rs
use crate::error::Result;
use crate::models::ChatMetadata;
use rusqlite::Connection;
use serde_json::Value;
use std::collections::HashMap;

/// Get all chat IDs from the database
pub fn get_all_chat_ids(conn: &Connection) -> Result<Vec<String>> {
    let mut stmt = conn.prepare("SELECT key FROM cursorDiskKV WHERE key LIKE 'composerData:%'")?;

    let chat_ids: Vec<String> = stmt
        .query_map([], |row| {
            let key: String = row.get(0)?;
            Ok(key.replace("composerData:", ""))
        })?
        .collect::<std::result::Result<Vec<String>, rusqlite::Error>>()?;

    Ok(chat_ids)
}

/// Get metadata for a specific chat
pub fn get_chat_metadata(conn: &Connection, chat_id: &str) -> Result<Option<ChatMetadata>> {
    let key = format!("composerData:{}", chat_id);
    let mut stmt = conn.prepare("SELECT value FROM cursorDiskKV WHERE key = ?")?;

    let result = stmt.query_row([&key], |row| {
        let value_str: String = row.get(0)?;
        Ok(value_str)
    });

    match result {
        Ok(value_str) => {
            let metadata_json: HashMap<String, Value> = serde_json::from_str(&value_str)?;

            // Parse the metadata into our ChatMetadata struct
            let chat_metadata = ChatMetadata {
                chat_id: chat_id.to_string(),
                name: metadata_json
                    .get("name")
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string()),
                created_at: metadata_json.get("createdAt").and_then(|v| v.as_i64()),
                created_at_iso: None, // Computed later if needed
                last_updated_at: metadata_json.get("lastUpdatedAt").and_then(|v| v.as_i64()),
                last_updated_at_iso: None, // Computed later if needed
                is_archived: metadata_json
                    .get("isArchived")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(false),
                is_draft: metadata_json
                    .get("isDraft")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(false),
                total_lines_added: metadata_json
                    .get("totalLinesAdded")
                    .and_then(|v| v.as_i64())
                    .map(|n| n as i32)
                    .unwrap_or(0),
                total_lines_removed: metadata_json
                    .get("totalLinesRemoved")
                    .and_then(|v| v.as_i64())
                    .map(|n| n as i32)
                    .unwrap_or(0),
                subtitle: metadata_json
                    .get("subtitle")
                    .and_then(|v| v.as_str())
                    .map(|s| s.to_string()),
                message_count: 0, // Computed separately
            };

            Ok(Some(chat_metadata))
        }
        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
        Err(e) => Err(e.into()),
    }
}

/// Get all messages for a specific chat
pub fn get_chat_messages(
    conn: &Connection,
    chat_id: &str,
) -> Result<Vec<(String, HashMap<String, Value>)>> {
    let pattern = format!("bubbleId:{}:%", chat_id);
    let mut stmt =
        conn.prepare("SELECT key, value FROM cursorDiskKV WHERE key LIKE ? ORDER BY key")?;

    let messages: Result<Vec<(String, HashMap<String, Value>)>> = stmt
        .query_map([pattern], |row| {
            let key: String = row.get(0)?;
            let value_str: String = row.get(1)?;
            Ok((key, value_str))
        })?
        .map(|result| {
            result.map_err(|e| e.into()).and_then(|(key, value_str)| {
                let bubble_id = key.split(':').last().unwrap_or("").to_string();
                let bubble_data: HashMap<String, Value> = serde_json::from_str(&value_str)?;
                Ok((bubble_id, bubble_data))
            })
        })
        .collect();

    messages
}

/// Count total number of chats
pub fn count_chats(conn: &Connection, include_archived: bool) -> Result<i32> {
    if include_archived {
        let count: i32 = conn.query_row(
            "SELECT COUNT(*) FROM cursorDiskKV WHERE key LIKE 'composerData:%'",
            [],
            |row| row.get(0),
        )?;
        Ok(count)
    } else {
        let mut stmt =
            conn.prepare("SELECT value FROM cursorDiskKV WHERE key LIKE 'composerData:%'")?;

        let mut count = 0;
        let rows = stmt.query_map([], |row| {
            let value_str: String = row.get(0)?;
            Ok(value_str)
        })?;

        for row_result in rows {
            let value_str = row_result?;
            if let Ok(metadata) = serde_json::from_str::<HashMap<String, Value>>(&value_str) {
                if !metadata
                    .get("isArchived")
                    .and_then(|v| v.as_bool())
                    .unwrap_or(false)
                {
                    count += 1;
                }
            }
        }

        Ok(count)
    }
}
